src/Core/Content/Mail/Service/MailService.php

```diff
@@ -319,37 +318,36 @@ private function getMediaUrls(array $data, Context $context): array
         return $urls;
     }
 
-    private function getSalesChannelDomainCriteria(string $salesChannelId, Context $context): Criteria
-    {
-        $criteria = new Criteria([$salesChannelId]);
-        $criteria->setTitle('mail-service::resolve-sales-channel-domain');
-        $criteria->addAssociation('mailHeaderFooter');
-        $criteria->getAssociation('domains')
-            ->addFilter(
-                new EqualsFilter('languageId', $context->getLanguageId())
-            );
-
-        return $criteria;
-    }
-
     /**
      * @param array<string, mixed> $data
+     * @param array<string, mixed> $templateData
      */
-    private function isTestMode(array $data = []): bool
+    private function getSalesChannel(array $data, array $templateData, Context $context): ?SalesChannelEntity
     {
-        $stagingMode = $this->systemConfigService->getBool(SetupStagingEvent::CONFIG_FLAG);
-        if ($stagingMode) {
-            return true;
+        $salesChannel = $templateData['salesChannel'] ?? null;
+        if ($salesChannel instanceof SalesChannelEntity) {
+            return $salesChannel;
         }
 
-        return !empty($data['testMode']);
-    }
+        $salesChannelId = $data['salesChannelId'] ?? null;
+        if (\is_string($salesChannelId)) {
+            $criteria = new Criteria([$salesChannelId]);
+            $criteria->setTitle('mail-service::resolve-sales-channel-domain');
+            $criteria->addAssociation('mailHeaderFooter');
+            $criteria->getAssociation('domains')
+                ->addFilter(
+                    new EqualsFilter('languageId', $context->getLanguageId())
+                );
+
+            $salesChannel = $this->salesChannelRepository->search(
+                $criteria,
+                $context
+            )->getEntities()->first();
+
+            // Should never happen since we check in the validation that if a salesChannelId is present it is valid...
+            \assert($salesChannel !== null);
```

Comment:
why do we need that? the method allows to return null, so why not return directly the result of "first"?